packages/vue-components/src/panels/PanelBase.js

```diff
@@ -95,80 +95,133 @@ export default {
     },
     open() {
+      this.localMinimized = false;
+      this.localExpanded = true;
       this.wasRetrieverLoaded = true;
+      /*
+        After setting minimized to false / wasRetrieverLoaded to true,
+        we have to wait for DOM update (nextTick) before setting maxHeight for transition.
+      */
       this.$nextTick(() => {
-        this.localExpanded = true;
+        this.$refs.panel.style.maxHeight = `${this.$refs.panel.scrollHeight}px`;
       });
-      this.localMinimized = false;
     },
     openPopup() {
       window.open(this.popupUrl);
     },
-    setMaxHeight() {
-      // Don't play the transition for this case as the loading should feel 'instant'
+    retrieverUpdateMaxHeight() {
+      // src has finished loaded -- we set this flag to true so our event listener can set maxHeight to none
+      this.isRetrieverLoadDone = true;
+
+      if (this.preloadBool && !this.wasRetrieverLoaded) {
+        // Only preload, do not expand the panel.
+        return;
+      }
+      // Don't play the transition for this case as the loading should feel 'instant'.
       if (this.expandedBool) {
         this.$refs.panel.style.maxHeight = 'none';
         return;
       }
 
-      /*
-      Otherwise, since the vue transition is dependent on localExpanded, we have to manually
-      set our own transition end handlers here for the initial loading of the content.
-      */
-      const onExpandDone = () => {
-        this.$refs.panel.style.maxHeight = 'none';
-        this.$refs.panel.removeEventListener('transitionend', onExpandDone);
-      };
-
-      this.$refs.panel.addEventListener('transitionend', onExpandDone);
+      // For expansion transition to 'continue' after src is loaded.
       this.$refs.panel.style.maxHeight = `${this.$refs.panel.scrollHeight}px`;
     },
-    beforeExpand(el) {
-      el.style.maxHeight = '0';
-    },
-    duringExpand(el) {
-      jQuery('html').stop();
-      el.style.maxHeight = `${el.scrollHeight}px`;
-    },
-    afterExpand(el) {
-      el.style.maxHeight = 'none';
-    },
-    beforeCollapse(el) {
-      el.style.maxHeight = `${el.scrollHeight}px`;
+    initCollapsedPanelHeight() {
+      // To set panel's default collapsed panel height.
+      this.collapsedPanelHeight = 0;
     },
-    duringCollapse(el) {
-      if (this.$el.getBoundingClientRect().top < 0) {
-        jQuery('html').animate({
-          scrollTop: window.scrollY + this.$el.getBoundingClientRect().top - 3,
-        }, 500, 'swing');
+    initPanel() {
+      this.$refs.panel.addEventListener('transitionend', () => {
+        /*
+          If the panel's content grows, its height should accomodate it accordingly (e.g. nested panels).
+          We will always set the panel's maxHeight to 'none' after expansion.
+          This is to prevent maxHeight (which is used for transition) from restricting panel's growth.
+
+          At the end of first expansion transition, if src has not finished loading, then we should not
```

Comment:
Let's explain the high level rationale instead, and leave the reader to figure out the details from the code easily :-)

```
However, if the panel has a 'src' attribute and it has not finished loading,
then this is delegated later to the retriever src-loaded event handler,
allowing a second maxHeight transition once it is loaded.
```